File: BCA/Java/BinarySearch.java
public class BinarySearch {
  static int search(int[] set, int item) {
    int found = -1, left = 0, right = set.length - 1, mid = 0;
    while (left <= right) {
      mid = (left + right) / 2;
      if (set[mid] == item) {
        found = mid;
        break;
      } else if (item > set[mid]) {
        left = mid + 1;
      } else {
        right = mid - 1;
      }
    }
    return found;
  }

  public static void sort(int[] arr) {
    for (int i = 0; i < arr.length; i++) {
      for (int j = 0; j < arr.length - 1; j++)
        if (arr[j] > arr[j + 1]) {
          int tmp = arr[j];
          arr[j] = arr[j + 1];
          arr[j + 1] = tmp;
        }
    }
  }

  public static void dump(int[] arr) {
    for (int i = 0; i < arr.length; i++) {
      System.out.print(arr[i] + " ");
    }
  }

  public static void main(String[] args) {
    int[] arr = { 4, 6, 9, 1, 2, 5, 3, 8, 7 };
    System.out.print("Array:\t\t");
    dump(arr);
    sort(arr);
    System.out.print("\nSorted array:\t");
    dump(arr);
    int find = 7;
    int loc = search(arr, find);
    if (loc > -1) {
      System.out.println("\nOccurrence of " + arr[loc] + " is at index #" + loc);
    } else {
      System.out.println("\nCouldn't find " + find);
    }
  }
}

File: BCA/DataStructures/Binary Search.c
#include <stdio.h>
#include <time.h>
#include <stdlib.h>
#define MAX 100

int search(int set[], int length, int item) {
	int low = 0, high = length, mid, found = 0;
	while (!found) {
		if (set[high] < item) {
			mid = -1;
			found = 1;
			break;
		}
		mid = low + (high - low) / 2;
		if (set[mid] < item) low = mid + 1;
		if (set[mid] > item) high = mid - 1;
		if (set[mid] == item) found = 1;
	}
	return mid - 1;
}

void dump(int set[], int length) {
	printf("Set = { ");
	for (int i = 1; i <= length; i++)
		printf("%s%d%s", 
			   set[i] <= 9 ? " " : "", 
			   set[i], i >= length ? "":", ");
	printf(" }\n");
}

int main() {
	int set[MAX];
	printf("Enter number of elements (MAX %d): ", MAX);
	int n;
	scanf("%d", &n);
	srand(time(NULL));
	set[0] = 0;
	for (int i = 1; i <= n; i++) 
		set[i] = (rand() % 50) + set[i - 1];
	dump(set, n);
	while (1) {
		printf("Enter element to find: ");
		int find;
		scanf("%d", &find);
		int loc = search(set, n, find);
		if (loc < 0) printf("%d not found in current set.\n", find);
		else printf("%d is located at Set[%d].\n", find, loc);
	}
	return 0;
}

